slapo/sharding/shard_ops.py

```diff
@@ -1,187 +1,611 @@
+       sharded (usually due to a mistake in the user-written schedule), then
+       the tied parameters will be broken.
 
     Parameters
     ----------
-    module: torch.nn.Module
-        The module to shard.
-    param_name: str
-        The name of the parameter to shard.
-    sharded_size: int
-        The size of the sharded dimension.
-    axis: int
-        The axis to shard on.
+    sch : Schedule
+        The schedule of the module.
+    old_param : nn.Parameter
+        The existing parameter before sharding.
+    new_tensor : torch.Tensor
+        The sharded tensor.
 
     Returns
     -------
-    tuple
-        The output type and the axis (if the output is partitioned).
-        The output type will be None if the output inference function is not
-        registered.
+    nn.Parameter
+        The new parameter with the sharded tensor, or the already created
+        tied parameter.
     """
-    module_cls = module.__class__
-    if module_cls in POSTPROC_FN:
-        return POSTPROC_FN[module_cls](module, param_name, sharded_size, axis)
-    return None, None
+    if old_param in sch.metadata.tie_weights:
+        if id(sch.metadata.tie_weights[old_param]) != id(old_param):
+            # This parameter is tied to another parameter, and the other
+            # parameter is already sharded. In this case we directly
+            # register the sharded parameter to the module to keep them tied.
+            if new_tensor.shape != sch.metadata.tie_weights[old_param].shape:
+                raise RuntimeError(
+                    f"Parameter in {sch.path} is tied, "
+                    "but they have different sharded shapes: "
+                    f"{new_tensor.shape} vs "
+                    f"{sch.metadata.tie_weights[old_param].shape}"
+                )
+            new_param = sch.metadata.tie_weights[old_param]
+        else:
+            # The first parameter in this tie group is sharded.
+            new_param = nn.Parameter(new_tensor)
+            sch.metadata.tie_weights[old_param] = new_param
+    else:
+        new_param = nn.Parameter(new_tensor)
+    return new_param
 
 
-@register_postproc_shard_fn(nn.Linear)
-def _linear(module, param_name, sharded_size, axis):
-    """Update attributes and return the output type for linear layer.
-    It adjusts the input or output feature size to reflect the shard size,
-    and returns the output type (partial or partition) after sharding.
+class ShardMethod:
+    """The class with default sharding methods. Note that this class
+    should never be instantiated, because all methods are static methods.
+    """
 
-    Parameters
-    ----------
-    module: torch.nn.Linear
-        The linear layer to shard.
-    param_name: str
-        The name of the parameter to shard.
-    sharded_size: int
-        The size of the sharded dimension.
-    axis: int
-        The axis to shard on.
+    @staticmethod
+    def preproc(sch, param_name, sharded_size, axis):
```

Comment:
`preproc` and `postproc` are for forward pass, then how can we handle the backward pass? Did you all reduce the bias gradient as what was done in #23?

Reply:
See (5) in description.